// @axelar-network/axelar-local-dev/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.mainnetInfo = exports.testnetInfo = exports.networks = exports.getDepositAddress = exports.stopAll = exports.stop = exports.setupNetwork = exports.getAllNetworks = exports.getNetwork = exports.forkNetwork = exports.createNetwork = exports.listen = exports.getGasPrice = exports.getFee = exports.contracts = void 0;
const info_1 = require("./info");
Object.defineProperty(exports, "mainnetInfo", { enumerable: true, get: function () { return info_1.mainnetInfo; } });
Object.defineProperty(exports, "testnetInfo", { enumerable: true, get: function () { return info_1.testnetInfo; } });
const Network_1 = require("./Network");
Object.defineProperty(exports, "networks", { enumerable: true, get: function () { return Network_1.networks; } });
const compiledContracts = __importStar(require("./contracts"));
const networkUtils_1 = require("./networkUtils");
Object.defineProperty(exports, "createNetwork", { enumerable: true, get: function () { return networkUtils_1.createNetwork; } });
Object.defineProperty(exports, "forkNetwork", { enumerable: true, get: function () { return networkUtils_1.forkNetwork; } });
Object.defineProperty(exports, "getAllNetworks", { enumerable: true, get: function () { return networkUtils_1.getAllNetworks; } });
Object.defineProperty(exports, "getDepositAddress", { enumerable: true, get: function () { return networkUtils_1.getDepositAddress; } });
Object.defineProperty(exports, "getFee", { enumerable: true, get: function () { return networkUtils_1.getFee; } });
Object.defineProperty(exports, "getGasPrice", { enumerable: true, get: function () { return networkUtils_1.getGasPrice; } });
Object.defineProperty(exports, "getNetwork", { enumerable: true, get: function () { return networkUtils_1.getNetwork; } });
Object.defineProperty(exports, "listen", { enumerable: true, get: function () { return networkUtils_1.listen; } });
Object.defineProperty(exports, "setupNetwork", { enumerable: true, get: function () { return networkUtils_1.setupNetwork; } });
Object.defineProperty(exports, "stop", { enumerable: true, get: function () { return networkUtils_1.stop; } });
Object.defineProperty(exports, "stopAll", { enumerable: true, get: function () { return networkUtils_1.stopAll; } });
const utils_1 = require("./utils");
__exportStar(require("./exportUtils"), exports);
__exportStar(require("./relay"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./Network"), exports);
exports.contracts = compiledContracts;
exports.utils = {
    deployContract: utils_1.deployContract,
    defaultAccounts: utils_1.defaultAccounts,
    setJSON: utils_1.setJSON,
    setLogger: utils_1.setLogger,
};
