// @axelar-network/axelar-local-dev/dist/types/factories/@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarExpressExecutable__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAxelarExpressExecutable__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "AlreadyExecuted",
        type: "error",
    },
    {
        inputs: [],
        name: "ExpressExecutorAlreadySet",
        type: "error",
    },
    {
        inputs: [],
        name: "InsufficientValue",
        type: "error",
    },
    {
        inputs: [],
        name: "InvalidAddress",
        type: "error",
    },
    {
        inputs: [],
        name: "NotApprovedByGateway",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        name: "ExpressExecuted",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                indexed: true,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                indexed: true,
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        name: "ExpressExecutedWithToken",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        name: "ExpressExecutionFulfilled",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                indexed: false,
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                indexed: true,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                indexed: true,
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        name: "ExpressExecutionWithTokenFulfilled",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
        ],
        name: "execute",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "string",
                name: "tokenSymbol",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "executeWithToken",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
        ],
        name: "expressExecute",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "expressExecuteWithToken",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [],
        name: "gateway",
        outputs: [
            {
                internalType: "contract IAxelarGateway",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
        ],
        name: "getExpressExecutor",
        outputs: [
            {
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commandId",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "sourceChain",
                type: "string",
            },
            {
                internalType: "string",
                name: "sourceAddress",
                type: "string",
            },
            {
                internalType: "bytes32",
                name: "payloadHash",
                type: "bytes32",
            },
            {
                internalType: "string",
                name: "symbol",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "getExpressExecutorWithToken",
        outputs: [
            {
                internalType: "address",
                name: "expressExecutor",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IAxelarExpressExecutable__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IAxelarExpressExecutable__factory = IAxelarExpressExecutable__factory;
IAxelarExpressExecutable__factory.abi = _abi;
