// @axelar-network/axelar-local-dev/dist/types/factories/@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IGovernable__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IGovernable__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "InvalidGovernance",
        type: "error",
    },
    {
        inputs: [],
        name: "InvalidMintLimiter",
        type: "error",
    },
    {
        inputs: [],
        name: "NotGovernance",
        type: "error",
    },
    {
        inputs: [],
        name: "NotMintLimiter",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousGovernance",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newGovernance",
                type: "address",
            },
        ],
        name: "GovernanceTransferred",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousGovernance",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newGovernance",
                type: "address",
            },
        ],
        name: "MintLimiterTransferred",
        type: "event",
    },
    {
        inputs: [],
        name: "governance",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "mintLimiter",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newGovernance",
                type: "address",
            },
        ],
        name: "transferGovernance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newGovernance",
                type: "address",
            },
        ],
        name: "transferMintLimiter",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IGovernable__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IGovernable__factory = IGovernable__factory;
IGovernable__factory.abi = _abi;
