// @axelar-network/axelar-local-dev/dist/types/factories/@axelar-network/axelar-cgp-solidity/contracts/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenDeployer__factory = exports.Ownable__factory = exports.MintableCappedERC20__factory = exports.EternalStorage__factory = exports.ERC20Permit__factory = exports.ERC20__factory = exports.ECDSA__factory = exports.DepositHandler__factory = exports.BurnableMintableCappedERC20__factory = exports.AxelarGatewayProxy__factory = exports.AxelarGateway__factory = exports.util = exports.interfaces = exports.gasService = exports.auth = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.auth = __importStar(require("./auth"));
exports.gasService = __importStar(require("./gas-service"));
exports.interfaces = __importStar(require("./interfaces"));
exports.util = __importStar(require("./util"));
var AxelarGateway__factory_1 = require("./AxelarGateway__factory");
Object.defineProperty(exports, "AxelarGateway__factory", { enumerable: true, get: function () { return AxelarGateway__factory_1.AxelarGateway__factory; } });
var AxelarGatewayProxy__factory_1 = require("./AxelarGatewayProxy__factory");
Object.defineProperty(exports, "AxelarGatewayProxy__factory", { enumerable: true, get: function () { return AxelarGatewayProxy__factory_1.AxelarGatewayProxy__factory; } });
var BurnableMintableCappedERC20__factory_1 = require("./BurnableMintableCappedERC20__factory");
Object.defineProperty(exports, "BurnableMintableCappedERC20__factory", { enumerable: true, get: function () { return BurnableMintableCappedERC20__factory_1.BurnableMintableCappedERC20__factory; } });
var DepositHandler__factory_1 = require("./DepositHandler__factory");
Object.defineProperty(exports, "DepositHandler__factory", { enumerable: true, get: function () { return DepositHandler__factory_1.DepositHandler__factory; } });
var ECDSA__factory_1 = require("./ECDSA__factory");
Object.defineProperty(exports, "ECDSA__factory", { enumerable: true, get: function () { return ECDSA__factory_1.ECDSA__factory; } });
var ERC20__factory_1 = require("./ERC20__factory");
Object.defineProperty(exports, "ERC20__factory", { enumerable: true, get: function () { return ERC20__factory_1.ERC20__factory; } });
var ERC20Permit__factory_1 = require("./ERC20Permit__factory");
Object.defineProperty(exports, "ERC20Permit__factory", { enumerable: true, get: function () { return ERC20Permit__factory_1.ERC20Permit__factory; } });
var EternalStorage__factory_1 = require("./EternalStorage__factory");
Object.defineProperty(exports, "EternalStorage__factory", { enumerable: true, get: function () { return EternalStorage__factory_1.EternalStorage__factory; } });
var MintableCappedERC20__factory_1 = require("./MintableCappedERC20__factory");
Object.defineProperty(exports, "MintableCappedERC20__factory", { enumerable: true, get: function () { return MintableCappedERC20__factory_1.MintableCappedERC20__factory; } });
var Ownable__factory_1 = require("./Ownable__factory");
Object.defineProperty(exports, "Ownable__factory", { enumerable: true, get: function () { return Ownable__factory_1.Ownable__factory; } });
var TokenDeployer__factory_1 = require("./TokenDeployer__factory");
Object.defineProperty(exports, "TokenDeployer__factory", { enumerable: true, get: function () { return TokenDeployer__factory_1.TokenDeployer__factory; } });
